### PR TITLE
feat(repository): Add to settings artifact repository

### PR DESCRIPTION
**Pull request by fleXible**
I'm reopening this PR, because #303 seems to be dead. I just fixed the tests, updated the docs a bit and resolved all merge conflicts to current master.

I use GitLab myself and am looking for an integrated workflow between it's package repositories, release feature and pipeline automation.

@relekang Maybe you can have a look at it. Feedback is welcome

### CODE REVIEW

File: semantic_release/repository.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+from semantic_release.settings import config
+
+from .helpers import LoggedFunction
+
+logger = logging.getLogger(__name__)
+
+
+class Repository(object):
+
+    def __init__(self, username: str = None, password: str = None) -> None:
```
Comment (relekang) [inline semantic_release/repository.py:20]: Make these not optional, having the default spread between this and the get_repository method can be confusing.

File: semantic_release/repository.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+from semantic_release import ImproperConfigurationError
+
+from semantic_release.settings import config
+
+from .helpers import LoggedFunction
+
+logger = logging.getLogger(__name__)
+
+
+class Repository(object):
```
Comment (relekang) [inline semantic_release/repository.py:18]: What is the benefit of turning the function that was here into a class?

**Review by relekang — CHANGES_REQUESTED** [review]

File: semantic_release/repository.py (hunk 2 above)
Comment (fleXible) [reply semantic_release/repository.py:18]: As stated in the PR, I'm not the original author of most of the code, so I can just take a guess. The original author is Benjamin René from #303 
My contributions are documentation, cleanups & fixing the failed tests.

Comment (relekang) [reply semantic_release/repository.py:18]: Oh, sorry 🙈

Comment (fleXible) [reply semantic_release/repository.py:161]: No worries, I made some progress last weekend but still needs some cleaning up before I can put it up. At least it's working now, but still far from perfect.
Still I just pushed it to get some feedback.

File: semantic_release/cli.py [hunk 3]
```diff
@@ -344,6 +332,31 @@ def publish(**kwargs):
     # else: Since version shows a message on failure, we do not need to print another.
 
 
+def repo(**kwargs: Dict[str, Any]):
```
Comment (fleXible) [inline semantic_release/cli.py:335]: Added that to quickly be able to do integration testing.

**Comment by fleXible** [comment]
Okay, I think it's done. Tests should all be passing, at least they do locally on my system.

**Comment by fleXible** [comment]
Obviously "should be passing" can be quite dramatically different from "are passing".
I had another go at it, now it shall be better. I couldn't fix the 2nd commitlint error, as it refers to a commit that's now burried deep in the history so rewording landed me in rebase-hell. And secondly it's not even a commit I authored.

**Comment by relekang** [comment]
Thanks for taking the time to finish this 😄

**Comment by fleXible** [comment]
> Thanks for taking the time to finish this 😄

To be honest, it was not easy, but I learned a lot in the process. 
And it's always good to see some code working the way I imagined it in the beginning!

Btw. I've found the "common options" not beeing common at all. Like some commands do have no-op handling, while others don't, or only partially.

**Comment by relekang** [comment]
Yeah, I think this codebase at the moment is suffering a bit from a lot of small fixes without a common direction breaking things like the common options. I think they used to work at some point 🤷

**Comment by fleXible** [comment]
> Yeah, I think this codebase at the moment is suffering a bit from a lot of small fixes without a common direction breaking things like the common options. I think they used to work at some point 🤷

I find the whole "common options" quite confusing. A big part of that comes from the fact, that   "--major", "--minor", "--patch" and "--retry" are separate flags, but they exclude each other.
Also the effect of "--post" with commands other than "changelog" is unclear, probably has no effect. "publish" will post the changelog by default with no way to disable it.

**Pull request merged**